app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx

```diff
@@ -0,0 +1,263 @@
+import { h, Fragment } from 'preact';
+import { useState, useEffect, useRef, useLayoutEffect } from 'preact/hooks';
+import PropTypes from 'prop-types';
+import {
+  Combobox,
+  ComboboxInput,
+  ComboboxPopover,
+  ComboboxList,
+  ComboboxOption,
+} from '@reach/combobox';
+import '@reach/combobox/styles.css';
+import { getMentionWordData, getCursorXY } from '@utilities/textAreaUtils';
+import { useMediaQuery, BREAKPOINTS } from '@components/useMediaQuery';
+
+const MIN_SEARCH_CHARACTERS = 2;
+const MAX_RESULTS_DISPLAYED = 6;
+
+const UserListItemContent = ({ user }) => {
+  return (
+    <Fragment>
+      <span className="crayons-avatar crayons-avatar--l mr-2 shrink-0">
+        <img
+          src={user.profile_image_90}
+          alt=""
+          className="crayons-avatar__image "
+        />
+      </span>
+
+      <div>
+        <p className="crayons-autocomplete__name">{user.name}</p>
+        <p className="crayons-autocomplete__username">{`@${user.username}`}</p>
+      </div>
+    </Fragment>
+  );
+};
+
+/**
+ * A component for dynamically searching for users and displaying results in a dropdown.
+ * This component will replace the textarea passed in props, copying all styles and attributes, and allowing for progressive enhancement
+ *
+ * @param {object} props
+ * @param {element} props.replaceElement The textarea DOM element that should be replaced
+ * @param {function} props.fetchSuggestions The async call to use for the search
+ *
+ * @example
+ * <MentionAutocompleteCombobox
+ *    replaceElement={textAreaRef.current}
+ *    fetchSuggestions={fetchUsersByUsername}
+ * />
+ */
+export const MentionAutocompleteTextArea = ({
+  replaceElement,
+  fetchSuggestions,
+}) => {
+  const [textContent, setTextContent] = useState('');
+  const [searchTerm, setSearchTerm] = useState('');
+  const [cachedSearches, setCachedSearches] = useState({});
+  const [dropdownPositionPoints, setDropdownPositionPoints] = useState({
+    x: 0,
+    y: 0,
+  });
+  const [selectionInsertIndex, setSelectionInsertIndex] = useState(0);
+  const [users, setUsers] = useState([]);
+  const [cursorPosition, setCursorPosition] = useState(null);
+  const [ariaHelperText, setAriaHelperText] = useState('');
+
+  const isSmallScreen = useMediaQuery(`(max-width: ${BREAKPOINTS.Small}px)`);
+
+  const inputRef = useRef(null);
+
+  useEffect(() => {
+    if (searchTerm.length >= MIN_SEARCH_CHARACTERS) {
+      if (cachedSearches[searchTerm]) {
```

Comment:
how long does this cache last? It's unlikely that it'll be an issue but I'm thinking of the scenario of a person that can't autocomplete a user's name because it didn't appear in the first page of results.

Reply:
It will last for as long as the user is on the current page and the component's unmounted. My thinking is that the first 6 results aren't going to change from the backend for a given search term in that time (I guess it is possible a new user registers or changes their username while we're on a page, but it seems unlikely we'd be searching for them in that case).

If the user adds another character to further refine the results, then we still end up with a new search call so I feel like this is OK